### PR TITLE
v0.5.0 (Beet PD3)

### PR DESCRIPTION
**Pull request by clockworkgr**
v0.5.0 CHANGE LOG

- Renamed everything to Beet from BTS Companion
- Removed socket.io dependency and wrote custom websocket handler (this was required for future mobile implementation)
- Changed static websocket client connection to instance-based to allow for multiple instances/connections
- Hardened comms security:

API calls are AES encrypted using a HOTP one-time-pass derived from a shared secret (established upon linking an app using ECDH with randomised keypairs) and a random nonce unique to each request. Each request also sets the hash of the next request nonce to provide forward security.

- Enabled SSL
- Implemented app/domain storage on IndexedDB using Dexie

THIS PR IS FOR COMPATIBILITY WITH THE CHANGES IN bitshares/beet#9
Time/Effort will be logged at that issue in the Beet repo.

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+
+
+
+
+
+
+
+
+
+class Beet {
```
Comment (sschiessl-bcp) [inline src/index.js]: Please add documentation to the head, at least what all the possible flags mean

File: src/index.js [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+            } catch (e) {
+                reject(false);
+            }
+        });
+    }
+    isInstalled() {
+        return new Promise(resolve => {
+            setTimeout(() => {
+                resolve(false);
+            }, 3000);
```
Comment (sschiessl-bcp) [inline src/index.js]: 3 second is too long. It will be resolving on localhost, I suggest something < 500ms

File: src/index.js [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+            this.ping().then(found => {
+                if (found) resolve(found);
+            });
+
+        })
+    }
+    async get_id() {
```
Comment (sschiessl-bcp) [inline src/index.js]: Always a random id? Maybe call it create_id then

File: src/index.js [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+        let new_id = await this.next_id();
+        let next_hash = await CryptoJS.SHA256('' + new_id).toString();
+        return {
+            id: id,
+            next_hash: next_hash.toString()
+        };
+    }
+    async next_id() {
+        if (this.connected && this.authenticated && this.linked) {
+            let next_id = Math.round(Math.random() * 100000000 + 1);
```
Comment (sschiessl-bcp) [inline src/index.js]: Not using get_id() ?

File: src/index.js [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+            await BeetClientDB.apps.where("apphash").equals(this.identity.apphash).modify({
+                next_id: next_id
+            });
+            return next_id;
+        } else {
+            throw new Error("You must be connected, authorised and linked.");
+        }
+    }
+    async link(chain = 'BTS') {
```
Comment (sschiessl-bcp) [inline src/index.js]: Please describe what link does

File: src/index.js [hunk 6]
```diff
@@ -0,0 +1,324 @@
...
+            if (!this.connected) throw new Error("You must connect to Beet first.");
+            if (!this.initialised) throw new Error("You must initialise the Beet Client first via init(appName).");            
+            setTimeout(() => {
+                resolve(false);
+            }, this.options.linkTimeout);
+            this.privk = crypto.randomBytes(32);
+            let pubkey = await eccrypto.getPublic(this.privk).toString('hex');
+            this.secret = await eccrypto.derive(this.privk, Buffer.from(this.beetkey, 'hex'));
+            var next_id = Math.round(Math.random() * 100000000 + 1);
+            this.chain = "BTS";
```
Comment (sschiessl-bcp) [inline src/index.js]: BTS hardcoded? Not using argument?

File: src/index.js [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+                    next_id: next_id
+                });
+                this.authenticated = res.authenticate;
+                this.linked = res.link;
+                this.identity = await BeetClientDB.apps.where("apphash").equals(this.apphash).first();
+
+                this.otp = new OTPAuth.HOTP({
+                    issuer: "Beet",
+                    label: "BeetAuth",
+                    algorithm: "SHA1",
```
Comment (sschiessl-bcp) [inline src/index.js]: Why not SHA256?

File: src/index.js [hunk 8]
```diff
@@ -0,0 +1,324 @@
...
+                    secret: OTPAuth.Secret.fromHex(this.identity.secret)
+                });
+                resolve(res);
+            }).catch(rej => {
+                reject(rej);
+            });
+        });
+
+    }
+    async connect(identity = null, options) {
```
Comment (sschiessl-bcp) [inline src/index.js]: Docs

File: src/index.js [hunk 9]
```diff
@@ -0,0 +1,324 @@
...
+                var auth = this.sendRequest('authenticate', authobj);
+                auth.then(res => {
+                    console.log(res);
+                    this.authenticated = res.authenticate;
+                    this.linked = res.link;
+                    if (this.linked) {
+                        this.otp = new OTPAuth.HOTP({
+                            issuer: "Beet",
+                            label: "BeetAuth",
+                            algorithm: "SHA1",
```
Comment (sschiessl-bcp) [inline src/index.js]: SHA256?

File: src/index.js (hunk 1 above)
Comment (clockworkgr) [reply src/index.js]: ok. will do

File: src/index.js (hunk 4 above)
Comment (clockworkgr) [reply src/index.js]: good point....I refactored that part a few times so I didn't notice

File: src/index.js (hunk 7 above)
Comment (clockworkgr) [reply src/index.js]: It was the default and there is no real reason to change it as it's a one-time hash and not the primary authentication mechanism (see: https://www.quora.com/Why-is-the-SHA1-algorithm-still-being-used-with-2FA-codes-instead-of-SHA2)

**Pull request merged**